pkgs/development/beam-modules/build-mix.nix

```diff
@@ -1,100 +1,70 @@
-{ stdenv, writeText, elixir, erlang, hex, lib }:
-
-{ name
-, version
-, src
-, setupHook ? null
-, buildInputs ? []
-, beamDeps ? []
-, postPatch ? ""
-, compilePorts ? false
-, installPhase ? null
-, buildPhase ? null
-, configurePhase ? null
-, meta ? {}
-, enableDebugInfo ? false
-, ... }@attrs:
-
-with lib;
+{ stdenv, lib, elixir, erlang, hex, rebar, rebar3, fetchMixDeps }:
 
-let
+{ name, version, src, nativeBuildInputs ? [ ], meta ? { }
+, enableDebugInfo ? false, depsSha256, mixEnv ? "prod", ... }@attrs:
 
-  debugInfoFlag = lib.optionalString (enableDebugInfo || elixir.debugInfo) "--debug-info";
+let
 
-  shell = drv: stdenv.mkDerivation {
-          name = "interactive-shell-${drv.name}";
-          buildInputs = [ drv ];
+  shell = drv:
+    stdenv.mkDerivation {
+      name = "interactive-shell-${drv.name}";
+      buildInputs = [ drv ];
     };
 
-  bootstrapper = ./mix-bootstrap;
+  mixDeps = fetchMixDeps {
+    inherit src name mixEnv version;
+    sha256 = depsSha256;
+  };
 
-  pkg = self: stdenv.mkDerivation ( attrs // {
-    name = "${name}-${version}";
-    inherit version;
+  pkg = self:
+    stdenv.mkDerivation (attrs // {
+      name = "${name}-${version}";
+      inherit version;
 
-    dontStrip = true;
+      dontStrip = true;
 
-    inherit src;
+      inherit src;
 
-    setupHook = if setupHook == null
-    then writeText "setupHook.sh" ''
-       addToSearchPath ERL_LIBS "$1/lib/erlang/lib"
-    ''
-    else setupHook;
+      nativeBuildInputs = nativeBuildInputs ++ [ erlang hex elixir ];
 
-    inherit buildInputs;
-    propagatedBuildInputs = [ hex elixir ] ++ beamDeps;
+      MIX_ENV = mixEnv;
+      MIX_DEBUG = if enableDebugInfo then 1 else 0;
+      HEX_OFFLINE = 1;
+      DEBUG = if enableDebugInfo then 1 else 0; # for Rebar3 compilation
 
-    configurePhase = if configurePhase == null
-    then ''
-      runHook preConfigure
-      ${erlang}/bin/escript ${bootstrapper}
-      runHook postConfigure
-    ''
-    else configurePhase ;
+      postUnpack = ''
+        export HEX_HOME="$TMPDIR/hex"
+        export MIX_HOME="$TMPDIR/mix"
+        export MIX_DEPS_PATH="$TMPDIR/deps"
 
+        # Rebar
+        # the api with `mix local.rebar rebar path` makes a copy of the binary
+        export MIX_REBAR "${rebar}/bin/rebar"
+        export MIX_REBAR3 "${rebar3}/bin/rebar3"
```

Comment:
you're right, I pushed a little too late last night. Let me fix that real quick so that at least it's testable. I'll take care of the other comments afterwards. Thanks a lot for all the comments! Super helpful!